ui/ozone/platform/wayland/wayland_nested_compositor.cc

```diff
@@ -0,0 +1,364 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ui/ozone/platform/wayland/wayland_nested_compositor.h"
+
+#include <wayland-client-protocol.h>
+#include <wayland-server-core.h>
+
+#include "base/files/file_path.h"
+#include "base/memory/ptr_util.h"
+#include "ui/gl/gl_context_egl.h"
+#include "ui/gl/gl_egl_api_implementation.h"
+#include "ui/gl/gl_gl_api_implementation.h"
+#include "ui/gl/gl_surface_egl.h"
+#include "ui/ozone/common/egl_util.h"
+#include "ui/ozone/platform/wayland/wayland_connection.h"
+#include "ui/ozone/platform/wayland/wayland_nested_surface.h"
+#include "ui/ozone/platform/wayland/wayland_window.h"
+
+namespace ui {
+
+namespace {
+
+// Default wayland socket name.
+constexpr base::FilePath::CharType kSocketName[] =
+    FILE_PATH_LITERAL("chormium-wayland-nested-compositor");
+
+#if !defined(PFNEGLBINDWAYLANDDISPLAYWL)
+typedef EGLBoolean (*PFNEGLBINDWAYLANDDISPLAYWL)(EGLDisplay,
+                                                 struct wl_display*);
+#endif
+
+#if !defined(PFNEGLUNBINDWAYLANDDISPLAYWL)
+typedef EGLBoolean (*PFNEGLUNBINDWAYLANDDISPLAYWL)(EGLDisplay,
+                                                   struct wl_display*);
+#endif
+
+#if !defined(PFNEGLQUERYWAYLANDBUFFERWL)
+typedef EGLBoolean (*PFNEGLQUERYWAYLANDBUFFERWL)(EGLDisplay,
+                                                 struct wl_resource*,
+                                                 EGLint attribute,
+                                                 EGLint* value);
+#endif
+
+#if !defined(PFNEGLCREATEIMAGEKHRPROC)
+typedef EGLImageKHR (*PFNEGLCREATEIMAGEKHRPROC)(EGLDisplay,
+                                                EGLContext,
+                                                EGLenum target,
+                                                EGLClientBuffer,
+                                                const EGLint* attribList);
+#endif
+
+#if !defined(PFNEGLDESTROYIMAGEKHRPROC)
+typedef EGLBoolean (*PFNEGLDESTROYIMAGEKHRPROC)(EGLDisplay, EGLImageKHR);
+#endif
+
+#if !defined(PFNGLEGLIMAGETARGETTEXTURE2DOESPROC)
+typedef void (*PFNGLEGLIMAGETARGETTEXTURE2DOESPROC)(GLenum target,
+                                                    GLeglImageOES);
+#endif
+
+#if !defined(PFNEGLCREATEWAYLANDBUFFERFROMIMAGEWL)
+typedef struct wl_buffer*(EGLAPIENTRYP PFNEGLCREATEWAYLANDBUFFERFROMIMAGEWL)(
+    EGLDisplay dpy,
+    EGLImageKHR image);
+#endif
+
+#ifndef EGL_WAYLAND_BUFFER_WL
+#define EGL_WAYLAND_BUFFER_WL 0x31D5 /* eglCreateImageKHR target */
+#endif
+
+static PFNEGLBINDWAYLANDDISPLAYWL bind_display;
+static PFNEGLUNBINDWAYLANDDISPLAYWL unbind_display;
+static PFNEGLQUERYWAYLANDBUFFERWL query_buffer;
+static PFNEGLCREATEIMAGEKHRPROC create_image;
+static PFNEGLDESTROYIMAGEKHRPROC destroy_image;
+static PFNEGLCREATEWAYLANDBUFFERFROMIMAGEWL create_wayland_buffer_from_image;
+
+template <class T>
```

Comment:
maybe add a todo here, so that exo's server code can reuse it (IIRC, these functions are there too)?

Reply:
do you mean templates?

Reply:
right. there is some duplication on the templates here and exo's server.cc . I mean maybe add a todo to factor things out and reuse them both.

Reply:
done